lib/runtime/user-vm.js

```diff
@@ -26,36 +26,43 @@ exports.create = async function({userId, staticTerrainData, staticTerrainDataSiz
     if(!vms[userId]) {
         let isolate = new ivm.Isolate({snapshot, memoryLimit: 256 + staticTerrainDataSize/1024/1024});
         let context = await isolate.createContext();
-        await context.global.set('global', context.global.derefInto());
-        await context.global.set('_ivm', ivm);
-        await context.global.set('_isolate', isolate);
-        await context.global.set('_context', context);
-        await context.global.set('_worldSize', index.getWorldSize());
-
-        let nativeModInstance = await nativeMod.create(context);
-        await context.global.set('_nativeMod', nativeModInstance.derefInto());
-
-        await (await isolate.compileScript('_init();')).run(context);
-
-        let evalFn = await context.global.get('_evalFn');
-        let start = await context.global.get('_start');
-
-        await Promise.all(Object.keys(staticTerrainData).map(async (room) => {
-            await (await context.global.get('_setStaticTerrainData')).apply(undefined,
-                [room, new ivm.ExternalCopy(staticTerrainData[room]).copyInto()]);
-        }));
-
-        await (await isolate.compileScript('new ' + function () {
-            delete global._ivm;
-            delete global._isolate;
-            delete global._context;
-            delete global._init;
-            delete global._evalFn;
-            delete global._start;
-            delete global._setStaticTerrainData;
-            delete global._worldSize;
-            delete global._nativeMod;
-        })).run(context);
+        let [ nativeModInstance, initScript, cleanupScript ] = await Promise.all([
+            nativeMod.create(context),
+            isolate.compileScript('_init();'),
+            isolate.compileScript('new ' + function () {
+                delete global._ivm;
+                delete global._isolate;
+                delete global._context;
+                delete global._init;
+                delete global._evalFn;
+                delete global._start;
+                delete global._setStaticTerrainData;
+                delete global._worldSize;
+                delete global._nativeMod;
+            }),
+        ]);
+        await Promise.all([
+            context.global.set('global', context.global.derefInto()),
+            context.global.set('_ivm', ivm),
+            context.global.set('_isolate', isolate),
+            context.global.set('_context', context),
+            context.global.set('_worldSize', index.getWorldSize()),
+            context.global.set('_nativeMod', nativeModInstance.derefInto()),
+            initScript.run(context),
```

Comment:
`_init()` depends on `_isolate`, `_ivm` and other globals. Can we be sure that promises on lines 45-50 are already fulfilled when `initScript.run()` is executed?

Reply:
Yup. The JS spec guarantees that expressions in an array will be evaluated from left to right and isolated-vm guarantees that promises will be resolved in the order they are scheduled.